hypervisor/arch/x86/guest/vm.c

```diff
@@ -556,19 +581,28 @@ int32_t shutdown_vm(struct acrn_vm *vm)
 	if (vm->state == VM_PAUSED) {
 		vm->state = VM_POWERED_OFF;
 
+		mask = lapic_pt_enabled_pcpu_bitmap(vm);
+
 		foreach_vcpu(i, vm, vcpu) {
 			reset_vcpu(vcpu);
 			offline_vcpu(vcpu);
 
-			if (is_lapic_pt_enabled(vcpu)) {
-				bitmap_set_nolock(vcpu->pcpu_id, &mask);
+			if (bitmap_test(vcpu->pcpu_id, &mask)) {
+				/*
+				 * If the current pcpu needs to offline itself,
+				 * it will be done after shutdown_vm() completes
+				 * in the idle thread.
+				 */
+				if (get_pcpu_id() == vcpu->pcpu_id) {
+					bitmap_clear_nolock(vcpu->pcpu_id, &mask);
+				}
```

Comment:
I realized a mistake, we still need to loop VCPUs here to do make_pcpu_offline.
So we have to either 1) roll back to original way, or 2) add a new API to do make_pcpus_offline(mask).

You decide which one is best...

Reply:
Thanks for the feedback. I think the original code is fine since it's a pretty straightforward change. Keeping make_pcpu_offline() clean and simple seems like a good approach to me.

Reply:
OK, can u change the code back?

Reply:
Please take a look at the latest patch a636ac789931ff041dd25c14af550d94e85466d3. I took the self-pcpu case out of the for loop.
Or would you like me to remove lapic_pt_enabled_pcpu_bitmap()?